networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp

```diff
@@ -0,0 +1,362 @@
+// networkit-format
+
+#include <algorithm>
+#include <cassert>
+#include <omp.h>
+#include <queue>
+
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/centrality/GroupClosenessLocalSwaps.hpp>
+
+namespace NetworKit {
+
+GroupClosenessLocalSwaps::GroupClosenessLocalSwaps(const Graph &G, const std::vector<node> &group,
+                                                   count maxSwaps)
+    : GroupClosenessLocalSwaps(G, group.begin(), group.end(), maxSwaps) {}
+
+void GroupClosenessLocalSwaps::init() {
+    const auto n = G->upperNodeIdBound();
+
+    distance.assign(n, 0);
+    gamma.assign(n * group.size(), false);
+    canSwap.assign(group.size(), false);
+    visited.assign(n, false);
+
+    idxMap.clear();
+    idxMap.reserve(group.size());
+
+    stack.assign(n, 0);
+    farness.assign(group.size(), 0);
+    farnessDecrement.assign(group.size(), 0);
+    sumOfMins.assign(n, 0);
+
+    intDistributions.resize(omp_get_max_threads());
+    urngs.reserve(omp_get_max_threads());
+    for (index i = 0; i < static_cast<index>(omp_get_max_threads()); ++i)
+        urngs.emplace_back(Aux::Random::getURNG());
+
+    for (size_t i = 0; i < group.size(); ++i) {
+        const auto u = group[i];
+        idxMap[u] = i;
+        gamma[u * group.size() + i] = 1;
+    }
+
+#ifdef __AVX2__
+    randVec.resize(n);
+#else
+    randVec.resize(K * n);
+#endif // __AVX2__
+
+    totalSwaps = 0;
+    hasRun = false;
+}
+
+void GroupClosenessLocalSwaps::run() {
+    init();
+
+    while (findAndSwap() && ++totalSwaps < maxSwaps) {
+    }
+
+    hasRun = true;
+}
+
+bool GroupClosenessLocalSwaps::findAndSwap() {
+    bfsFromGroup();
+    const auto v = estimateHighestDecrement();
+    const auto farnessDecrementV = computeFarnessDecrement(v);
+    int64_t improvement = 0;
+    node u = none;
+
+    G->forNeighborsOf(v, [&](const node y) {
+        if (distance[y] == 0) {
+            const auto idx = idxMap.at(y);
+            const auto curImprovement = farnessDecrementV - farness[idx] + farnessDecrement[idx];
+
+            if (curImprovement > improvement) {
+                improvement = curImprovement;
+                u = y;
+            }
+        }
+    });
+
+    for (const auto &idx : idxMap) {
+        if (canSwap[idx.second]) {
+            int64_t curImprovement =
+                farnessDecrementV - farness[idx.second] + farnessDecrement[idx.second];
+            if (curImprovement > improvement) {
+                improvement = curImprovement;
+                u = idx.first;
+            }
+        }
+    }
```

Comment:
Why do we need this second loop? What does it do? Maybe add a comment.

Reply:
That's not needed, I probably forgot to remove it from a previous version. I removed it together with the `canSwap` vector, which was used just here.